envi-tools/src/TIFFIO.cpp

```diff
@@ -0,0 +1,115 @@
+void tio::WriteTIFF(const boost::filesystem::path& path, const cv::Mat& img)
+{
+    // Safety checks
+    if (img.channels() != 1) {
+        throw std::runtime_error("Multichannel images not supported");
+    }
+
+    auto ext = path.extension().string();
+    boost::to_upper(ext);
+    if (ext != ".TIF" && path.extension() != ".TIFF") {
+        throw std::runtime_error("Invalid file extension " + ext);
+    }
+
+    // Image metadata
+    auto channels = img.channels();
+    auto width = static_cast<unsigned>(img.cols);
+    auto height = static_cast<unsigned>(img.rows);
+    auto rowsPerStrip = height;
+
+    // Sample format
+    int bitsPerSample;
+    int sampleFormat;
+    switch (img.depth()) {
+        case CV_8U:
+            sampleFormat = SAMPLEFORMAT_UINT;
+            bitsPerSample = 8;
+            break;
+        case CV_8S:
+            sampleFormat = SAMPLEFORMAT_INT;
+            bitsPerSample = 8;
+            break;
+        case CV_16U:
+            sampleFormat = SAMPLEFORMAT_UINT;
+            bitsPerSample = 16;
+            break;
+        case CV_16S:
+            sampleFormat = SAMPLEFORMAT_INT;
+            bitsPerSample = 16;
+            break;
+        case CV_32F:
+            sampleFormat = SAMPLEFORMAT_IEEEFP;
+            bitsPerSample = 32;
+            break;
+        case CV_64F:
+            sampleFormat = SAMPLEFORMAT_IEEEFP;
+            bitsPerSample = 64;
+            break;
+        default:
+            throw std::runtime_error("Unsupported image depth");
+    }
+
+    // Photometric Interpretation
+    int photometric;
+    switch (channels) {
+        case 1:
+            photometric = PHOTOMETRIC_MINISBLACK;
+            break;
+        case 3:
+            photometric = PHOTOMETRIC_RGB;
+            break;
+        default:
+            throw std::runtime_error("Unsupported number of channels");
+    }
+
+    // Open the file
+    auto out = lt::TIFFOpen(path.c_str(), "w");
+    if (out == nullptr) {
+        throw std::runtime_error("Failed to open file for writing");
+    }
+
+    // Encoding parameters
+    lt::TIFFSetField(out, TIFFTAG_IMAGEWIDTH, width);
+    lt::TIFFSetField(out, TIFFTAG_IMAGELENGTH, height);
+    lt::TIFFSetField(out, TIFFTAG_PHOTOMETRIC, photometric);
+    lt::TIFFSetField(out, TIFFTAG_PLANARCONFIG, PLANARCONFIG_CONTIG);
+    lt::TIFFSetField(out, TIFFTAG_COMPRESSION, COMPRESSION_LZW);
+    lt::TIFFSetField(out, TIFFTAG_SAMPLEFORMAT, sampleFormat);
+    lt::TIFFSetField(out, TIFFTAG_BITSPERSAMPLE, bitsPerSample);
+    lt::TIFFSetField(out, TIFFTAG_SAMPLESPERPIXEL, channels);
+    lt::TIFFSetField(out, TIFFTAG_ROWSPERSTRIP, rowsPerStrip);
+
+    // Row buffer. OpenCV documentation mentions that TIFFWriteScanline
+    // modifies its read buffer, so we can't use the cv::Mat directly
+    auto bufferSize = static_cast<size_t>(lt::TIFFScanlineSize(out));
+    std::vector<char> buffer(bufferSize + 32);
+
+    // For each row
+    for (unsigned row = 0; row < height; row++) {
```

Comment:
It does default to an int. It's determined by the lhs of the assignment, in this case the 0.